### PR TITLE
Remove remaining experimental Kotlin language flags from the Kotlin DSL

### PR DESCRIPTION
**Pull request by bamboo**

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test this

**Review by eskatos — COMMENTED** [review]
It's great we can finally get rid of all experimental kotlin compiler flags \o/

Setting `kotlinDslPluginOptions.experimentalWarning` should emit a deprecation warning.
The internal task logger decoration mechanism can now be dropped.

Also, as discussed, the compiler flags could be moved to the provider so that we don't need a new plugin version when we change them. That would allow sharing them with the script compilation.

**Comment by bamboo** [comment]
@bot-gradle test RFN

**Comment by eskatos** [comment]
One thing I forgot in my comment above, the enabled experimental flags should now be removed from `subprojects/docs/src/docs/userguide/api/kotlin_dsl.adoc`.
See https://docs.gradle.org/nightly/userguide/kotlin_dsl.html#sec:kotlin_compiler_arguments

**Comment by bamboo** [comment]
> One thing I forgot in my comment above, the enabled experimental flags should now be removed from `subprojects/docs/src/docs/userguide/api/kotlin_dsl.adoc`.
> See https://docs.gradle.org/nightly/userguide/kotlin_dsl.html#sec:kotlin_compiler_arguments

Thanks for the reminder, @eskatos!

**Comment by bamboo** [comment]
@bot-gradle test RFN

**Comment by bamboo** [comment]
PTAL, @eskatos.

File: subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginOptions.kt [hunk 1]
```diff
@@ -43,9 +43,8 @@ class KotlinDslPluginOptions internal constructor(objects: ObjectFactory) {
     /**
      * Set to `false` to silence the warning about the `kotlin-dsl` plugin enabling Kotlin compiler experimental features.
      */
-    val experimentalWarning = objects.property<Boolean>().apply {
-        set(true)
-    }
+    @Deprecated("Flag has no effect since `kotlin-dsl` no longer relies on experimental features.")
```
Comment (eskatos) [inline subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginOptions.kt:46]: :x: This should have a removal version and nag.

File: subprojects/kotlin-dsl-plugins/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginTest.kt [hunk 2]
```diff
@@ -263,7 +262,7 @@ class KotlinDslPluginTest : AbstractPluginTest() {
 
     @Test
     @ToBeFixedForConfigurationCache(because = "Kotlin Gradle Plugin")
-    fun `by default experimental Kotlin compiler features are enabled and a warning is issued`() {
+    fun `can use SAM conversions for Kotlin functions without warnings`() {
```
Comment (eskatos) [inline subprojects/kotlin-dsl-plugins/src/integTest/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginTest.kt:265]: 👍

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinDslPluginSupport.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.provider
+
+
+object KotlinDslPluginSupport {
+
+    val kotlinCompilerArgs: List<String>
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinDslPluginSupport.kt:22]: 👍

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt [hunk 4]
```diff
@@ -340,12 +339,6 @@ private
 val gradleKotlinDslLanguageVersionSettings = LanguageVersionSettingsImpl(
     languageVersion = LanguageVersion.KOTLIN_1_4,
     apiVersion = ApiVersion.KOTLIN_1_4,
-    specificFeatures = mapOf(
-        LanguageFeature.DisableCompatibilityModeForNewInference to LanguageFeature.State.ENABLED,
-        LanguageFeature.SamConversionForKotlinFunctions to LanguageFeature.State.ENABLED,
-        LanguageFeature.SamConversionPerArgument to LanguageFeature.State.ENABLED,
-        LanguageFeature.ReferencesToSyntheticJavaProperties to LanguageFeature.State.ENABLED
-    ),
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt:348]: \o/

**Review by eskatos — CHANGES_REQUESTED** [review]

File: subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginOptions.kt (hunk 1 above)
Comment (bamboo) [reply subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginOptions.kt:48]: Done.

**Comment by bamboo** [comment]
@bot-gradle test this

File: subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginOptions.kt [hunk 5]
```diff
@@ -41,10 +43,24 @@ class KotlinDslPluginOptions internal constructor(objects: ObjectFactory) {
...
+            nagUserAboutExperimentalWarning()
+            return experimentalWarningProperty
+        }
+
+    private
+    val experimentalWarningProperty = objects.property<Boolean>()
+
+    private
+    fun nagUserAboutExperimentalWarning() {
+        deprecate("kotlinDslPluginOptions.experimentalWarning")
```
Comment (eskatos) [inline subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginOptions.kt:60]: :x: This declares a deprecated _feature_. It should use `deprecateProperty(name)` instead.

**Review by eskatos — CHANGES_REQUESTED** [review]

File: subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginOptions.kt (hunk 5 above)
Comment (eskatos) [reply subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/dsl/KotlinDslPluginOptions.kt]: I went ahead and refined the deprecation message

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
This will need a new plugin version

**Review by eskatos — CHANGES_REQUESTED** [review]
This will need a new plugin version before merging

**Comment by bamboo** [comment]
> This will need a new plugin version before merging

Could you do that? My setup is not ready for it.

**Comment by bamboo** [comment]
@bot-gradle test RFN

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Comment by eskatos** [comment]
@bamboo plugins published and dependency verification metadata updated. Last step will be a wrapper update and version bump in `build-logic-commons/gradle-plugin/build.gradle.kts`. That can happen either in this PR or afterwards.

**Review by eskatos — APPROVED** [review]

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Comment by bamboo** [comment]
Thanks, @eskatos! I believe the next steps will be more efficient from master.

**Pull request merged**